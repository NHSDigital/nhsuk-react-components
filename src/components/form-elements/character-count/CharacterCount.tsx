'use client';
import React, { FC, useEffect } from 'react';
import CharacterCountJs from '@resources/character-count';
import { HTMLAttributesWithData } from '@util/types/NHSUKTypes';

export enum CharacterCountType {
  Characters,
  Words,
}

type CharacterCountProps = React.HTMLAttributes<HTMLDivElement> & {
  children: React.ReactNode;
  maxLength: number;
  countType: CharacterCountType;
  textAreaId: string;
  thresholdPercent?: number;
};

const CharacterCount: FC<CharacterCountProps> = ({
  children,
  maxLength,
  countType,
  textAreaId,
  thresholdPercent,
  ...rest
}) => {
  useEffect(() => {
    CharacterCountJs();
  }, []);

  const characterCountProps: HTMLAttributesWithData<HTMLDivElement> =
    countType === CharacterCountType.Characters
      ? { ...rest, ['data-maxlength']: maxLength }
      : { ...rest, ['data-maxwords']: maxLength };

  if (thresholdPercent) {
    characterCountProps['data-threshold'] = thresholdPercent;
  }

  return (
    <div
      className="nhsuk-character-count"
      data-module="nhsuk-character-count"
      {...characterCountProps}
    >
      <div className="nhsuk-form-group">{children}</div>

      <div className="nhsuk-hint nhsuk-character-count__message" id={`${textAreaId}-info`}>
        You can enter up to {maxLength} characters
      </div>
    </div>
  );
};

export default CharacterCount;
