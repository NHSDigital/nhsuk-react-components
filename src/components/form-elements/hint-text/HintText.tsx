import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';

export type HintTextProps = HTMLProps<HTMLDivElement>;

const HintText: FC<HintTextProps> = ({ className, ...rest }) => (
  <div className={classNames('nhsuk-hint', className)} {...rest} />
);

export default HintText;
