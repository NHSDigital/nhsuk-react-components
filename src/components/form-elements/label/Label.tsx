import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';
import { NHSUKSize } from '@util/types/NHSUKTypes';

export interface LabelProps extends Omit<HTMLProps<HTMLLabelElement>, 'size'> {
  bold?: boolean;
  isPageHeading?: boolean;
  size?: NHSUKSize;
}

const BaseLabel: FC<LabelProps> = ({ className, bold, size, isPageHeading, ...rest }) => (
  // eslint-disable-next-line jsx-a11y/label-has-associated-control
  <label
    className={classNames(
      'nhsuk-label',
      { 'nhsuk-label--s': bold && !size },
      { 'nhsuk-label--xl': isPageHeading && !size },
      { [`nhsuk-label--${size}`]: size },
      className,
    )}
    {...rest}
  />
);

const Label: FC<LabelProps> = ({ isPageHeading, ...rest }) => {
  if (isPageHeading) {
    return (
      <h1 className="nhsuk-label-wrapper">
        <BaseLabel isPageHeading {...rest} />
      </h1>
    );
  }
  return <BaseLabel {...rest} />;
};

export default Label;
