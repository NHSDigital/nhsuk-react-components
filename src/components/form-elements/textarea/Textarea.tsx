import React, { FC, HTMLProps, MutableRefObject } from 'react';
import classNames from 'classnames';
import { FormElementProps } from '@util/types/FormTypes';
import SingleInputFormGroup from '@components/utils/SingleInputFormGroup';

type TextareaProps = HTMLProps<HTMLTextAreaElement> &
  FormElementProps & { textareaRef?: MutableRefObject<HTMLTextAreaElement | null> };

const Textarea: FC<TextareaProps> = ({ children, ...props }) => (
  <SingleInputFormGroup<TextareaProps> inputType="textarea" {...props}>
    {({ className, error, textareaRef, ...rest }) => (
      <textarea
        className={classNames('nhsuk-textarea', { 'nhsuk-textarea--error': error }, className)}
        ref={textareaRef}
        {...rest}
      >
        {children}
      </textarea>
    )}
  </SingleInputFormGroup>
);

export default Textarea;
