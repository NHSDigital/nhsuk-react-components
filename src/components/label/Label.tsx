import React, { HTMLProps, useContext } from 'react';
import classNames from 'classnames';
import { NHSUKSize } from '../../util/types/NHSUKTypes';
import FormGroupContext from '../formgroup/FormGroupContext';

export interface LabelProps extends Omit<HTMLProps<HTMLLabelElement>, 'size'> {
  bold?: boolean;
  isPageHeading?: boolean;
  size?: NHSUKSize;
}

const BaseLabel: React.FC<LabelProps> = props => {
  const { id, className, bold, size, isPageHeading, ...rest } = props;
  const { inputID } = useContext(FormGroupContext);
  const labelId = props.id || `${inputID}--label` || undefined;

  return (
    // eslint-disable-next-line jsx-a11y/label-has-associated-control
    <label
      id={labelId}
      className={classNames(
        'nhsuk-label',
        { 'nhsuk-label--s': bold && !size },
        { 'nhsuk-label--xl': isPageHeading && !size },
        { [`nhsuk-label--${size}`]: size },
        className,
      )}
      {...rest}
    />
  );
};

const Label: React.FC<LabelProps> = ({ isPageHeading, ...rest }) => {
  if (isPageHeading) {
    return (
      <h1 className="nhsuk-label-wrapper">
        <BaseLabel isPageHeading {...rest} />
      </h1>
    );
  }
  return <BaseLabel {...rest} />;
};

export default Label;
