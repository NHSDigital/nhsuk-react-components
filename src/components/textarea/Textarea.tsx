import React, { HTMLProps, MutableRefObject } from 'react';
import classNames from 'classnames';
import { FormElementProps } from '../../util/types/FormTypes';
import FormGroup from '../../util/FormGroup';

type TextareaProps = HTMLProps<HTMLTextAreaElement> &
  FormElementProps & { textareaRef?: MutableRefObject<HTMLTextAreaElement | null> };

const Textarea: React.FC<TextareaProps> = (props) => (
  <FormGroup<TextareaProps> inputType="textarea" {...props}>
    {({ className, error, textareaRef, ...rest }) => (
      <textarea
        className={classNames('nhsuk-textarea', { 'nhsuk-textarea--error': error }, className)}
        ref={textareaRef}
        {...rest}
      />
    )}
  </FormGroup>
);

export default Textarea;
