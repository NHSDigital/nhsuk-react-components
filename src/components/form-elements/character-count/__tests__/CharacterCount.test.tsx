import React from 'react';
import { render } from '@testing-library/react';
import CharacterCount, { CharacterCountType } from '../CharacterCount';
import Label from '@components/form-elements/label/Label';
import HintText from '@components/form-elements/hint-text/HintText';
import Textarea from '@components/form-elements/textarea/Textarea';

describe('Character Count', () => {
  it('Matches snapshot', () => {
    const { container } = render(
      <CharacterCount
        maxLength={200}
        countType={CharacterCountType.Characters}
        textAreaId="more-detail"
      >
        <Label htmlFor="more-detail">Can you provide more detail?</Label>
        <HintText id="more-detail-hint">
          Do not include personal information like your name, date of birth or NHS number.
        </HintText>
        <Textarea
          id="more-detail"
          className="nhsuk-js-character-count"
          name="more-detail"
          aria-describedby="more-detail-hint"
          rows={5}
        />
      </CharacterCount>,
    );

    expect(container).toMatchSnapshot();
  });

  it('Sets the data-maxlength attribute when counting characters', () => {
    const { container } = render(
      <CharacterCount
        maxLength={200}
        countType={CharacterCountType.Characters}
        textAreaId="more-detail"
      >
        <div />
      </CharacterCount>,
    );

    expect(container.querySelector('.nhsuk-character-count')?.getAttribute('data-maxlength')).toBe(
      '200',
    );
    expect(
      container.querySelector('.nhsuk-character-count')?.getAttribute('data-maxwords'),
    ).toBeNull();
    expect(
      container.querySelector('.nhsuk-character-count')?.getAttribute('data-threshold'),
    ).toBeNull();
  });

  it('Sets the data-maxwords attribute when counting words', () => {
    const { container } = render(
      <CharacterCount maxLength={200} countType={CharacterCountType.Words} textAreaId="more-detail">
        <div />
      </CharacterCount>,
    );

    expect(container.querySelector('.nhsuk-character-count')?.getAttribute('data-maxwords')).toBe(
      '200',
    );
    expect(
      container.querySelector('.nhsuk-character-count')?.getAttribute('data-maxlength'),
    ).toBeNull();
    expect(
      container.querySelector('.nhsuk-character-count')?.getAttribute('data-threshold'),
    ).toBeNull();
  });

  it('Sets the data-threshold attribute when threshold is specified', () => {
    const { container } = render(
      <CharacterCount
        maxLength={200}
        countType={CharacterCountType.Characters}
        thresholdPercent={50}
        textAreaId="more-detail"
      >
        <div />
      </CharacterCount>,
    );

    expect(container.querySelector('.nhsuk-character-count')?.getAttribute('data-threshold')).toBe(
      '50',
    );
  });
});
