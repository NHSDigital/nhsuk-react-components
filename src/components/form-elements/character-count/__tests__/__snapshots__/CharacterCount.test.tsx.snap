// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Count Matches snapshot 1`] = `
<div>
  <div
    class="nhsuk-character-count"
    data-maxlength="200"
    data-module="nhsuk-character-count"
  >
    <div
      class="nhsuk-form-group"
    >
      <label
        class="nhsuk-label"
        for="more-detail"
      >
        Can you provide more detail?
      </label>
      <div
        class="nhsuk-hint"
        id="more-detail-hint"
      >
        Do not include personal information like your name, date of birth or NHS number.
      </div>
      <div
        class="nhsuk-form-group"
      >
        <textarea
          aria-describedby="more-detail-hint"
          class="nhsuk-textarea nhsuk-js-character-count"
          id="more-detail"
          name="more-detail"
          rows="5"
        />
        <div
          class="nhsuk-hint nhsuk-character-count__message nhsuk-u-visually-hidden"
          id="more-detail-info"
        >
          You can enter up to 
          200
           characters
        </div>
        <div
          aria-hidden="true"
          class="nhsuk-hint nhsuk-character-count__message nhsuk-character-count__status"
        >
          You have 200 characters remaining
        </div>
        <div
          aria-live="polite"
          class="nhsuk-character-count__sr-status nhsuk-u-visually-hidden"
        >
          You have 200 characters remaining
        </div>
      </div>
    </div>
  </div>
</div>
`;
